x-pack/plugins/embeddable_enhanced/public/drilldowns/url_drilldown/url_drilldown_scope.tsx

```diff
@@ -0,0 +1,322 @@
+): Promise<ValueClickTriggerEventScope> {
+  const negate = eventScopeInput.data.negate ?? false;
+  const { table, column: columnIndex, value } = await getSingleValue(
+    eventScopeInput.data.data,
+    deps,
+    opts
+  );
+  const column = table.columns[columnIndex];
+  return cleanEmptyKeys({
+    key: toPrimitiveOrUndefined(column?.meta?.aggConfigParams?.field) as string | undefined,
+    value: toPrimitiveOrUndefined(value) as string | number | boolean,
+    negate,
+  });
+}
+
+/**
+ * @remarks
+ * Difference between `event` and `context` variables, is that real `context` variables are available during drilldown creation (e.g. embeddable panel)
+ * `event` variables are mapped from trigger context. Since there is no trigger context during drilldown creation, we have to provide some _mock_ variables for validating and previewing the URL
+ */
+export function getMockEventScope([trigger]: UrlTrigger[]): UrlDrilldownEventScope {
+  if (trigger === SELECT_RANGE_TRIGGER) {
+    return {
+      key: '__testKey__',
+      from: new Date(Date.now() - 15 * 60 * 1000).toISOString(), // 15 minutes ago
+      to: new Date().toISOString(),
+    };
+  } else {
+    return {
+      key: '__testKey__',
+      value: '__testValue__',
+      negate: false,
+    };
+  }
+}
+
+function toPrimitiveOrUndefined(v: unknown): string | number | boolean | undefined {
+  if (typeof v === 'number' || typeof v === 'boolean' || typeof v === 'string') return v;
+  if (typeof v === 'object' && v instanceof Date) return v.toISOString();
+  if (typeof v === 'undefined' || v === null) return undefined;
+  return String(v);
+}
+
+function cleanEmptyKeys<T extends Record<string, any>>(obj: T): T {
+  Object.keys(obj).forEach((key) => {
+    if (obj[key] === undefined) {
+      delete obj[key];
+    }
+  });
+  return obj;
+}
+
+/**
+ * VALUE_CLICK_TRIGGER could have multiple data points
+ * Prompt user which data point to use in a drilldown
+ */
+async function getSingleValue(
+  data: ValueClickContext['data']['data'],
+  deps: { getOpenModal: () => Promise<OverlayStart['openModal']> },
+  opts: { allowPrompts: boolean } = { allowPrompts: false }
+): Promise<ValueClickContext['data']['data'][0]> {
+  data = data.filter(Boolean);
+  if (data.length === 0)
+    throw new Error(`[trigger = "VALUE_CLICK_TRIGGER"][getSingleValue] no value to pick from`);
+  if (data.length === 1) return Promise.resolve(data[0]);
+  if (!opts.allowPrompts) return Promise.resolve(data[0]);
+  return new Promise(async (resolve, reject) => {
+    const openModal = await deps.getOpenModal();
+    const overlay = openModal(
+      toMountPoint(
+        <GetSingleValuePopup
+          onCancel={() => overlay.close()}
+          onSubmit={(point) => {
+            if (point) {
+              resolve(point);
+            }
+            overlay.close();
+          }}
+          data={data}
+        />
+      )
+    );
+    overlay.onClose.then(() => reject());
+  });
+}
+
+function GetSingleValuePopup({
```

Comment:
Shall we type it as a React component, e.g. `GetSingleValuePopup: React.FC<...>`? And maybe have it in a separate file.

Reply:
In general, for me it is easier to navigate code in a single file, then jumping around multiple files and in this case. So I tend to not prematurely create new files. 

In this case I agree that this _should probably be_ in a separate file, but since I expect significant changes in this area soon (with all the extension points), I want to keep everything related to url drilldown scope implementation in 1 file (because it is going to be changed together). (I think less files the change touch: easier to refactor and review). 
Also it likely will go away soon if we decide we want: https://github.com/Dosant/kibana/pull/5/files